File: kotlin/Backtracking/FindAllPermutations.kt
fun findAllPermutations(nums: List<Int>): List<List<Int>> {
    val res = mutableListOf<List<Int>>()
    backtrack(nums, mutableListOf(), mutableSetOf(), res)
    return res
}

fun backtrack(
        nums: List<Int>,
        candidate: MutableList<Int>,
        used: MutableSet<Int>,
        res: MutableList<List<Int>>
) {
    // If the current candidate is a complete permutation, add it to the
    // result.
    if (candidate.size == nums.size) {
        res.add(candidate.toList())
        return
    }
    for (num in nums) {
        if (num !in used) {
            // Add 'num' to the current permutation and mark it as used.
            candidate.add(num)
            used.add(num)
            // Recursively explore all branches using the updated
            // permutation candidate.
            backtrack(nums, candidate, used, res)
            // Backtrack by reversing the changes made.
            candidate.removeAt(candidate.size - 1)
            used.remove(num)
        }
    }
}

File: cpp/Backtracking/find_all_permutations.cpp
#include <vector>
#include <set>

std::vector<std::vector<int>> findAllPermutations(std::vector<int>& nums) {
    std::vector<std::vector<int>> res;
    std::vector<int> candidate;
    std::set<int> used;
    backtrack(nums, candidate, used, res);
    return res;
}

void backtrack(std::vector<int>& nums, std::vector<int>& candidate, std::set<int>& used, std::vector<std::vector<int>>& res) {
    // If the current candidate is a complete permutation, add it to the 
    // result.
    if (candidate.size() == nums.size()) {
        res.push_back(candidate);
        return;
    }
    for (int num : nums) {
        if (used.find(num) == used.end()) {
            // Add 'num' to the current permutation and mark it as used.
            candidate.push_back(num);
            used.insert(num);
            // Recursively explore all branches using the updated 
            // permutation candidate.
            backtrack(nums, candidate, used, res);
            // Backtrack by reversing the changes made.
            candidate.pop_back();
            used.erase(num);
        }
    }
}